File: servo-motor.js
// Contrôle des moteurs SERVO42D NEMA17

const { PINS } = require('./esp32-pins');

class ServoMotor {
  constructor(motorId = 1, uart = null) {
    this.motorId = motorId;
    this.uart = uart; // Connexion UART sur RX/TX pins 16/17
    this.position = 0; // position en steps
    this.isEnabled = false;
    this.direction = 1; // 1 = forward, -1 = backward
    this.speed = 500; // steps/s
    this.maxSteps = 100000; // limite max
  }

  // Activer le moteur
  enable() {
    if (this.isEnabled) return this;

    this.isEnabled = true;
    // UART command: {motorId}:ENABLE
    this._sendCommand('ENABLE');
    console.log(`✅ Moteur ${this.motorId} ACTIVÉ`);
    return this;
  }

  // Désactiver le moteur
  disable() {
    if (!this.isEnabled) return this;

    this.isEnabled = false;
    this._sendCommand('DISABLE');
    console.log(`❌ Moteur ${this.motorId} DÉSACTIVÉ`);
    return this;
  }

  // Bouger d'un nombre de steps
  moveSteps(steps) {
    if (!this.isEnabled) {
      console.log(`⚠️ Moteur ${this.motorId} doit être activé d'abord`);
      return this;
    }

    const newPos = this.position + steps;
    if (newPos < 0 || newPos > this.maxSteps) {
      console.log(`⚠️ Position hors limites: ${newPos} (max: ${this.maxSteps})`);
      return this;
    }

    this.position = newPos;
    const direction = steps > 0 ? 1 : -1;
    this.direction = direction;

    // UART: {motorId}:MOVE:{steps}:{speed}
    this._sendCommand(`MOVE:${Math.abs(steps)}:${this.speed}`);
    console.log(`⬆️ Moteur ${this.motorId} -> ${Math.abs(steps)} steps (pos: ${this.position})`);
    return this;
  }

  // Aller à une position absolue
  goToPosition(targetPos) {
    if (targetPos < 0 || targetPos > this.maxSteps) {
      console.log(`⚠️ Position invalide: ${targetPos}`);
      return this;
    }

    const steps = targetPos - this.position;
    return this.moveSteps(steps);
  }

  // Réinitialiser à la position 0
  home() {
    const steps = -this.position;
    this.position = 0;
    this._sendCommand(`MOVE:${Math.abs(steps)}:${this.speed}`);
    console.log(`🏠 Moteur ${this.motorId} retour à HOME (pos: 0)`);
    return this;
  }

  // Définir la vitesse (steps/s)
  setSpeed(stepsPerSecond) {
    if (stepsPerSecond < 0) {
      console.log('⚠️ Vitesse doit être positive');
      return this;
    }

    this.speed = stepsPerSecond;
    this._sendCommand(`SPEED:${stepsPerSecond}`);
    console.log(`⚙️ Moteur ${this.motorId} vitesse -> ${stepsPerSecond} steps/s`);
    return this;
  }

  // Définir le couple/force
  setTorque(percent) {
    if (percent < 0 || percent > 100) {
      console.log('⚠️ Couple doit être entre 0 et 100%');
      return this;
    }

    this._sendCommand(`TORQUE:${percent}`);
    console.log(`💪 Moteur ${this.motorId} couple -> ${percent}%`);
    return this;
  }

  // Mode micro-stepping (1, 2, 4, 8, 16)
  setMicrostepping(division) {
    const valid = [1, 2, 4, 8, 16];
    if (!valid.includes(division)) {
      console.log(`⚠️ Micro-stepping invalide: ${division}`);
      return this;
    }

    this._sendCommand(`MICROSTEPPING:${division}`);
    console.log(`🔧 Moteur ${this.motorId} micro-stepping -> 1/${division}`);
    return this;
  }

  // Obtenir l'état
  getStatus() {
    return {
      motorId: this.motorId,
      isEnabled: this.isEnabled,
      position: this.position,
      direction: this.direction === 1 ? 'forward' : 'backward',
      speed: this.speed,
      maxSteps: this.maxSteps,
    };
  }

  // Afficher l'état
  printStatus() {
    const s = this.getStatus();
    console.log(`\n📊 État moteur ${s.motorId}:`);
    console.log(`   Statut: ${s.isEnabled ? '🟢 ACTIVÉ' : '🔴 DÉSACTIVÉ'}`);
    console.log(`   Position: ${s.position} steps`);
    console.log(`   Direction: ${s.direction}`);
    console.log(`   Vitesse: ${s.speed} steps/s`);
    return this;
  }

  // Envoyer une commande UART
  _sendCommand(cmd) {
    const message = `M${this.motorId}:${cmd}\n`;
    if (this.uart) {
      // Code réel ESP32: Serial1.print(message);
      // Pour tests:
      console.log(`📡 UART TX: ${message.trim()}`);
    } else {
      console.log(`📡 [SIM] UART TX: ${message.trim()}`);
    }
  }
}

// Classe pour gérer deux moteurs ensemble
class DualServoMotor {
  constructor(uart = null) {
    this.motor1 = new ServoMotor(1, uart);
    this.motor2 = new ServoMotor(2, uart);
  }

  // Activer les deux
  enableAll() {
    this.motor1.enable();
    this.motor2.enable();
    return this;
  }

  // Désactiver les deux
  disableAll() {
    this.motor1.disable();
    this.motor2.disable();
    return this;
  }

  // Bouger les deux en synchrone
  moveBothSteps(steps1, steps2) {
    this.motor1.moveSteps(steps1);
    this.motor2.moveSteps(steps2);
    return this;
  }

  // Mouvement différentiel (avant/arrière)
  differential(leftSteps, rightSteps) {
    this.motor1.moveSteps(leftSteps);
    this.motor2.moveSteps(rightSteps);
    console.log(`🔄 Mouvement différentiel: L:${leftSteps}, R:${rightSteps}`);
    return this;
  }

  // Tourner sur place
  turnInPlace(steps, direction = 1) {
    const leftSteps = direction > 0 ? steps : -steps;
    const rightSteps = direction > 0 ? -steps : steps;
    return this.differential(leftSteps, rightSteps);
  }

  getStatus() {
    return {
      motor1: this.motor1.getStatus(),
      motor2: this.motor2.getStatus(),
    };
  }

  printStatus() {
    this.motor1.printStatus();
    this.motor2.printStatus();
    return this;
  }
}

module.exports = { ServoMotor, DualServoMotor };

File: src/servo_motor.cpp
// Implémentation contrôle moteurs SERVO42D

#include "servo_motor.hpp"
#include "driver/uart.h"
#include <cstdio>
#include <cstring>

// UART pins pour ESP32
#define UART_TX_PIN 17
#define UART_RX_PIN 16
#define UART_NUM UART_NUM_1
#define UART_BAUD_RATE 115200
#define UART_BUF_SIZE 256

ServoMotor::ServoMotor(uint8_t id, int uart_num)
    : motor_id_(id), uart_num_(uart_num) {
    state_.motor_id = id;
    state_.position = 0;
    state_.enabled = false;
    state_.speed = 500;
    state_.torque = 100;
    state_.microstepping = 1;

    // Initialiser UART si premier moteur
    if (id == 1) {
        uart_config_t uart_config = {
            .baud_rate = UART_BAUD_RATE,
            .data_bits = UART_DATA_8_BITS,
            .parity = UART_PARITY_DISABLE,
            .stop_bits = UART_STOP_BITS_1,
            .flow_ctrl = UART_HW_FLOWCTRL_DISABLE,
        };

        uart_param_config(UART_NUM, &uart_config);
        uart_set_pin(UART_NUM, UART_TX_PIN, UART_RX_PIN,
                     UART_PIN_NO_CHANGE, UART_PIN_NO_CHANGE);
        uart_driver_install(UART_NUM, UART_BUF_SIZE * 2, UART_BUF_SIZE * 2, 0, nullptr, 0);

        printf("[ServoMotor] UART initialized on pins TX:%d RX:%d\n", UART_TX_PIN, UART_RX_PIN);
    }
}

ServoMotor::~ServoMotor() {
    if (state_.enabled) {
        disable();
    }
}

void ServoMotor::move_steps(int32_t steps) {
    if (!state_.enabled) {
        printf("[ServoMotor%d] Motor must be enabled first\n", motor_id_);
        return;
    }

    state_.position += steps;

    char cmd[64];
    snprintf(cmd, sizeof(cmd), "M%d:MOVE:%ld:%u\n",
             motor_id_, labs(steps), state_.speed);

    send_command(cmd);
    printf("[ServoMotor%d] Moved %ld steps (pos: %ld)\n", motor_id_, steps, state_.position);
}

void ServoMotor::go_to_position(int32_t target_pos) {
    if (target_pos < 0) {
        printf("[ServoMotor%d] Invalid position\n", motor_id_);
        return;
    }

    int32_t steps = target_pos - state_.position;
    move_steps(steps);
}

void ServoMotor::home() {
    move_steps(-state_.position);
    state_.position = 0;
    printf("[ServoMotor%d] Homed to position 0\n", motor_id_);
}

void ServoMotor::enable() {
    state_.enabled = true;
    send_command("ENABLE");
    printf("[ServoMotor%d] Enabled\n", motor_id_);
}

void ServoMotor::disable() {
    state_.enabled = false;
    send_command("DISABLE");
    printf("[ServoMotor%d] Disabled\n", motor_id_);
}

void ServoMotor::set_speed(uint32_t steps_per_second) {
    state_.speed = steps_per_second;

    char cmd[64];
    snprintf(cmd, sizeof(cmd), "M%d:SPEED:%u\n", motor_id_, steps_per_second);
    send_command(cmd);

    printf("[ServoMotor%d] Speed set to %u steps/s\n", motor_id_, steps_per_second);
}

void ServoMotor::set_torque(uint16_t percent) {
    if (percent > 100) percent = 100;

    state_.torque = percent;

    char cmd[64];
    snprintf(cmd, sizeof(cmd), "M%d:TORQUE:%u\n", motor_id_, percent);
    send_command(cmd);

    printf("[ServoMotor%d] Torque set to %u%%\n", motor_id_, percent);
}

void ServoMotor::set_microstepping(uint8_t division) {
    if (division != 1 && division != 2 && division != 4 &&
        division != 8 && division != 16) {
        printf("[ServoMotor%d] Invalid microstepping value\n", motor_id_);
        return;
    }

    state_.microstepping = division;

    char cmd[64];
    snprintf(cmd, sizeof(cmd), "M%d:MICROSTEPPING:%u\n", motor_id_, division);
    send_command(cmd);

    printf("[ServoMotor%d] Microstepping set to 1/%u\n", motor_id_, division);
}

ServoMotorState ServoMotor::get_status() const {
    return state_;
}

void ServoMotor::send_command(const std::string& cmd) {
    uart_write_bytes(UART_NUM, cmd.c_str(), cmd.length());
    printf("[UART TX] M%d: %s", motor_id_, cmd.c_str());
}

// ============= DualServoMotor =============

DualServoMotor::DualServoMotor(int uart_num)
    : motor1(1, uart_num), motor2(2, uart_num) {
}

void DualServoMotor::enable_all() {
    motor1.enable();
    motor2.enable();
}

void DualServoMotor::disable_all() {
    motor1.disable();
    motor2.disable();
}

void DualServoMotor::move_both(int32_t steps1, int32_t steps2) {
    motor1.move_steps(steps1);
    motor2.move_steps(steps2);
}

void DualServoMotor::differential(int32_t left, int32_t right) {
    printf("[DualServoMotor] Differential: L:%ld R:%ld\n", left, right);
    move_both(left, right);
}

void DualServoMotor::turn_in_place(int32_t steps, int direction) {
    int32_t left = direction > 0 ? steps : -steps;
    int32_t right = direction > 0 ? -steps : steps;

    printf("[DualServoMotor] Turning in place: %ld steps, direction: %d\n", steps, direction);
    differential(left, right);
}
